nemo_skills/evaluation/evaluator/math.py

```diff
@@ -58,30 +60,108 @@ class LeanEvaluatorConfig:
+    asyncio.run(evaluator.eval_full(cfg.input_files))
+
+
+# Evaluator Classes
+
+
+class MathEvaluator(BaseEvaluator):
+    """Math evaluator - only supports batch evaluation."""
+
+    async def eval_full(self, input_files: list[str], **kwargs) -> None:
+        """Batch evaluate math problems."""
+        eval_config = MathEvaluatorConfig(**self.config)
+        eval_config_dict = asdict(eval_config)
+        batch_evaluate_results(
+            input_files=input_files,
+            **eval_config_dict,
+        )
+
+
+class Lean4ProofEvaluator(BaseEvaluator):
+    """Lean4 proof evaluator - supports both single and batch evaluation."""
+
+    async def eval_full(self, input_files: list[str], **kwargs) -> None:
+        """Batch evaluate Lean4 proofs."""
+        eval_config = LeanEvaluatorConfig(**self.config)
+        sandbox = get_sandbox(**eval_config.sandbox)
+        eval_config_dict = asdict(eval_config)
+        eval_config_dict.pop("sandbox")
+        await sandbox.batch_evaluate_results(
+            input_files=input_files,
             answer_format="lean4-proof",
             **eval_config_dict,
         )
-    )
 
+    async def eval_single(self, data_point: dict[str, any]) -> dict[str, any]:
+        """Evaluate single Lean4 proof during generation."""
+        eval_config = LeanEvaluatorConfig(**self.config)
+        sandbox = get_sandbox(**eval_config.sandbox)
 
-def eval_lean4_statement(cfg):
-    eval_config = LeanEvaluatorConfig(**cfg.eval_config)
-
-    sandbox = get_sandbox(**eval_config.sandbox)
-    eval_config_dict = asdict(eval_config)
-    eval_config_dict.pop("sandbox")
-    asyncio.run(
-        sandbox.batch_evaluate_results(
-            input_files=cfg.input_files,
+        # Prepare predicted_proof using shared utility
+        generation = data_point["generation"]
+
+        config = ProofBuildConfig(
+            final_answer_key=eval_config.final_answer_key,
+            extract_code_mode=eval_config.extract_code_mode,
+            restate_formal_statement=eval_config.restate_formal_statement,
+            strip_theorem_from_proof=True,  # Default behavior for proofs
+        )
+
+        predicted_proof = build_lean4_proof(
+            generation=generation, data_point=data_point, config=config, answer_format="lean4-proof"
+        )
+
+        # Execute proof and get compiler output
+        try:
+            output, _ = await sandbox.execute_code(
+                generated_code=predicted_proof,
+                language="lean4",
+                timeout=eval_config.timeout,
+            )
+
+            # Determine proof status using shared utility
+            proof_status = determine_proof_status(output)
+
+            return {
+                "predicted_proof": predicted_proof,
+                "proof_status": proof_status,
+                "lean_evaluation": {**output, "timeout": eval_config.timeout},
+            }
+
+        except Exception as e:
```

Comment:
I'm not sure having blank exception is good here. I'd rather let it fail, so that users quickly see what's wrong. Otherwise maybe there is 10% of samples where we have some bug in eval that would be incorrectly evaluated and it's hard to notice unless you really look into the generation file and spot those